### PR TITLE
[SIP-4][superset-client] add `SupersetClient`, refactor app

### PR DESCRIPTION
**Pull request by williaster**
This PR implements [SIP-4](https://github.com/apache/incubator-superset/issues/5667), replacing all JS ajax calls with a new `SupersetClient` that handles all async requests using `fetch` 🎉  

It's a big PR due to the ajax refactoring, so here's an higher-altitude summary:

#### `SupersetClient` usage*

```javascript
// appSetup.js
import SupersetClient from `@superset-ui/core`;

SupersetClient.configure(...clientConfig);
SupersetClient.init(); // CSRF auth, can also chain `.configure().init();

// anotherFile.js
import SupersetClient from `@superset-ui/core`;

SupersetClient.post(...requestConfig)
  .then(({ request, json }) => ...)
  .catch((error) => ...);
```

*this is currently in `src/packages/core/src` which will become `@superset-ui/core`

#### `SupersetClient` handles:
- `CSRF` token authentication 
    - it also queues requests in the case that another request is made before the token is received
    - it checks for a token before every request, an external app that uses this can detect this by catching errors, or explicitly checking `SupersetClient.isAuthorized()`
- timeouts (not supported by default in fetch)
- query aborts (also not supported by default in fetch, uses `AbortController` polyfill)
- supports the following configuration options: 
    - `protocol = 'http'`
    - `host`
    - `headers`
    - `credentials = 'same-origin'` (`include` for non-Superset apps)
    - `mode = 'same-origin'`  (`cors` for non-Superset apps)
    - `timeout`
- supports `get`s and `post`s. I considered adding `put`s/`delete`s but no existing ajax calls of this type are currently made
- a given request supports the following options, which met the need of every async call in the app:
  - `url` or `endpoint`
  - `headers`
  - `body`
  - `timeout`
  - `signal` (for aborting, from `const { signal } = (new AbortController())`)
  - for posts
    - `postPayload` (key values are added to a `new FormData()`)
    - `stringify` whether to call `JSON.stringify` on `postPayload` values

#### New dependencies
It introduces the following new packages:
- `abortcontroller-polyfill`
- `whatwg-fetch`
- `fetch-mock` for testing

#### How was this tested?
- attempted to test success + failure of every endpoint that was refactored in the browser
- updated broken / ajax-containing tests
- (TODO) add `SupersetClient` tests 

#### Next steps
0. finish todos
1. merge this PR, test in staging, fix any immediate bugs
2. move `packages/core` to `superset-ui` repo
3. update imports to pull from `@superset-ui/core`

#### TODO
- [x] implement `SupersetClient`
- [x] refactor all ajax calls to use `SupersetClient` / `fetch`
- [x] refactor / fix all ajax-containing tests
- [x] debug SQL lab polling
- [x] remove unused `SupersetClient` methods like `put` and `destroy`
- [x] refactor API
- [ ] rebase
- [ ] add `SupersetClient` tests
- [ ] add API documentation to package readme
- [ ] verify sqllab error links

@mistercrunch @kristw @conglei @graceguo-supercat @hughhhh @betodealmeida @michellethomas

### CODE REVIEW

File: superset/assets/.eslintrc [hunk 1]
```diff
@@ -26,7 +27,7 @@
     "no-mixed-operators": 0,
     "no-continue": 0,
     "no-bitwise": 0,
-    "no-undef": 0,
+    "no-undef": 1,
```
Comment (williaster) [inline superset/assets/.eslintrc]: warning for this, turning this off is BAD!

File: superset/assets/package.json [hunk 2]
```diff
@@ -43,10 +43,12 @@
   },
   "homepage": "http://superset.apache.org/",
   "dependencies": {
-    "@data-ui/event-flow": "^0.0.54",
+    "@data-ui/event-flow": "^0.0.62",
```
Comment (williaster) [inline superset/assets/package.json]: will remove this, left from something else

File: superset/assets/.eslintrc (hunk 1 above)
Comment (kristw) [reply superset/assets/.eslintrc]: +100

Comment (kristw) [reply superset/assets/.eslintrc]: This one total worths its own PR.

**Comment by williaster** [comment]
Okay so there are some concerns that this PR is too large and so we're considering the following:

1) abandon this PR
2) create a PR in the superset-ui repo with just the `SupersetClient` package
3) merge then publish the package
4) create a series of smaller PRs for the ajax refactors, possibly adding integration tests to them as we go.

File: superset/assets/spec/javascripts/dashboard/reducers/sliceEntities_spec.js [hunk 3]
```diff
@@ -42,7 +42,7 @@ describe('sliceEntities reducer', () => {
       {},
       {
         type: FETCH_ALL_SLICES_FAILED,
-        error: { responseJSON: { message: 'errorrr' } },
+        error: 'errorrr',
```
[the remark this answers is not in the document]
Comment (williaster) [reply superset/assets/spec/javascripts/dashboard/reducers/sliceEntities_spec.js]: this is a test / misspelling was purposeful.

**Pull request closed without merging**